Documentation restructure (#251)
signalfx/signalfx-dotnet-tracing: docs/advanced-config.md

# Advanced Configuration

## Configuration Methods

> There are following ways to apply configuration settings

Reviewer: ```suggestion
You can apply configuration settings in the following ways
```

> (prioritized from top to bottom):

Reviewer: ```suggestion
, by order of preference:
```

1. Environment variables

>     Environment variables are the main way to configure values.
>     A setting configured via an environment variable cannot be overridden.

Reviewer: ```suggestion
Settings configured using environment variables always take precedence.
```

There are several ways of configuring env vars in Windows. Can we provide two sample snippets for PowerShell and Linux?

2. `web.config` or `app.config` file

>     For an application running on .NET Framework, web configuration file (`web.config`)
>     or application configuration file (`app.config`) can be used to configure settings.

Reviewer: ```suggestion
    For an application running on .NET Framework, you can use a web configuration file 
    (`web.config`) or an application configuration file (`app.config`) to configure settings.
```

    Example with `SIGNALFX_SERVICE` setting:

>     ```xml

Reviewer: Do XML settings take precedence over JSON config files?

Author: yes they do

    <configuration>
    <appSettings>
        <add key="SIGNALFX_SERVICE" value="my-service-name" />
    </appSettings>
    </configuration>
    ```

3. JSON configuration file

>     To use it, set environment variable `SIGNALFX_TRACE_CONFIG_FILE`

Reviewer: Any hint as to why would someone prefer this method over XML?

Author: I think that `app.config` and `web.config` is working OOTB only on .NET Framework. So XML is rather for .NET (Core) applications. @RassK @Kielek am I correct?

Reviewer 2: Currently I would say JSON is a backup option for .NET core stuff. It's easier to set local environment variables when running .net core,  so I guess env's will be still primary.

Author: The only scenario I see is when a .NET 5.0 app is deployed IIS. @RassK does it make sense?

Reviewer 2: I can't think of any necessary scenario atm. With .NET Core you can always scope envs locally, even when hosting in IIS. (I think newer IIS is still reusing the .NET Core feature to set envs).

Might be just for some kind of CI, that works better with JSON.

| `SIGNALFX_VERSION` | The application's version that will populate `version` tag on spans. |  |
| `SIGNALFX_TRACE_HEADER_TAGS` | Comma-separated map of header keys to tag name, that will be automatically applied as tags on traces. | `"key1:val1,key2:val2"` |
| `SIGNALFX_MAX_TRACES_PER_SECOND` | The number of traces allowed to be submitted per second. | `100` |
| `SIGNALFX_TRACE_RESPONSE_HEADER_ENABLED` | Enable to add server trace information to HTTP response headers. | `true` |
| `SIGNALFX_TRACE_STARTUP_LOGS` | Enable to activate diagnostic log at stratup. | `true` |
| `SIGNALFX_TRACE_SAMPLING_RULES` | Comma separated list of sampling rules taht enabled custom sampling rules based on regular expressions. The rule is matched in order of specification. The first match in a list is used. The item "sample_rate" is required in decimal format. The item "service" is optional in regular expression format, to match on service name. The item "name" is optional in regular expression format, to match on operation name. | `'[{"sample_rate":0.5, "service":"cart.*"}],[{"sample_rate":0.2, "name":"http.request"}]'` |

> | `SIGNALFX_TRACE_SAMPLE_RATE` | The global rate for the sampler. |  |

Reviewer: No default?

Author: When not assigned there then it is always sampling. I will try to address it in a separate PR.
